File: ios/HybridNitroDocumentPicker.swift
//
//  HybridNitroDocumentPicker.swift
//  NitroDocumentPicker
//
//  Created by Patrick Kabwe on 6/7/2025.
//

import Foundation
import NitroModules

class HybridNitroDocumentPicker: HybridNitroDocumentPickerSpec {
    private let nitroDocPickerImpl = NitroDocumentPickerImpl()

    func pick(options: NitroDocumentPickerOptions) throws -> Promise<Variant__NitroDocumentPickerResult__NitroDocumentPickerResult> {
        return .async { @MainActor [weak self] in
            guard let self = self else {
                throw RuntimeError.error(withMessage: "HybridNitroDocumentPicker instance has been deallocated")
            }
            if options.multiple == true {
                let results = try await self.nitroDocPickerImpl.pick(options: options)
                return Variant__NitroDocumentPickerResult__NitroDocumentPickerResult.first(results)
            }
            let results = try await self.nitroDocPickerImpl.pick(options: options)
            
            return Variant__NitroDocumentPickerResult__NitroDocumentPickerResult.second(results[0])
        }
    }
    
    func pickDirectory() throws -> Promise<NitroDocumentPickerDirectoryResult> {
        return .async { @MainActor [weak self] in
            guard let self = self else {
                throw RuntimeError.error(withMessage: "HybridNitroDocumentPicker instance has been deallocated")
            }
            return try await self.nitroDocPickerImpl.pickDirectory()
        }
    }
}

File: android/src/main/java/com/nitrodocumentpicker/HybridNitroDocumentPicker.kt
package com.nitrodocumentpicker

import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitrodocumentpicker.HybridNitroDocumentPickerSpec
import com.margelo.nitro.nitrodocumentpicker.NitroDocumentPickerOptions
import com.margelo.nitro.nitrodocumentpicker.NitroDocumentPickerDirectoryResult
import com.margelo.nitro.nitrodocumentpicker.Variant_Array_NitroDocumentPickerResult__NitroDocumentPickerResult

class HybridNitroDocumentPicker: HybridNitroDocumentPickerSpec() {
    val context = NitroModules.applicationContext ?: throw Exception("Context not found")
    val picker = NitroDocumentPicker(context)

    override fun pick(options: NitroDocumentPickerOptions): Promise<Variant_Array_NitroDocumentPickerResult__NitroDocumentPickerResult> {
        return Promise.async {
            try {
                val results = picker.pick(options)
                if (options.multiple == true) {
                    Variant_Array_NitroDocumentPickerResult__NitroDocumentPickerResult.First(results)
                } else {
                    Variant_Array_NitroDocumentPickerResult__NitroDocumentPickerResult.Second(results[0])
                }
            } catch (e: Exception) {
                e.printStackTrace()
                throw Error(e)
            }
        }
    }


    override fun pickDirectory(): Promise<NitroDocumentPickerDirectoryResult> {
        return Promise.async {
            try {
                picker.pickDirectory()
            } catch (e: Exception) {
                e.printStackTrace()
                throw Error(e)
            }
        }
    }
}
